CUDARelativeDifferencePrior

Imraj-Singh opened the pull request:
First attempt at integrating CudaRelativeDifferencePrior class into STIR. The idea was create a child of `RelativeDifferencePrior` to override the `compute_value`/`compute_gradient` methods with the CUDA-accelerated counterpart, and then add an `set_up_cuda` method. This is in a first-draft state with work left to do:
1. There are a lot of warnings with regards to `-- virtual function override intended?`.
2. Only a 3x3x3 neighbourhood size is implemented.
3. Threads and blocks sizes are not set well.
4. The set_up_cuda is a bit clunky (I did have it override the set-up previously but got a bit confused with CRTP things)

At present it does run with a test comparing outputs of CPU/GPU versions. I could include it in the pull-request but I need to spend sometime making the ctest I guess.

KrisThielemans commented on CMakeLists.txt:
```diff
@@ -207,6 +208,20 @@ else()
 endif()
 
+if(NOT DISABLE_STIR_CUDA)
+  find_package(CUDAToolkit)
+  if (CUDAToolkit_FOUND)
+    set(CMAKE_CUDA_ARCHITECTURES "all")
+    find_package(CUDA REQUIRED)
+    include_directories("${CUDA_INCLUDE_DIRS}")
+    set(STIR_WITH_STIR_CUDA ON)
+    if(STIR_WITH_STIR_CUDA)
+      enable_language(CUDA)
+      message(STATUS "STIR CUDA support enabled. Using CUDA version ${CUDAToolkit_VERSION}.")
+    endif()
+  endif()
```
I think this is the other way around. I'd probably use https://cmake.org/cmake/help/latest/module/CheckLanguage.html#module:CheckLanguage first, if present, then `enable_language`, then [`find_package(CUDAToolkit)`](https://cmake.org/cmake/help/latest/module/FindCUDAToolkit.html) which probably needs a `REQUIRED`.

Before all this, I'd do a check on CMAKE_VERSION. Certainly at least 3.18, but 3.23 if we want to use `"all" for 
[CMAKE_CUDA_ARCHITECTURES](https://cmake.org/cmake/help/latest/prop_tgt/CUDA_ARCHITECTURES.html#prop_tgt:CUDA_ARCHITECTURES) (which we do). If the CMake version is too old, generate FATAL_ERROR with the message to set `DISABLE_STIR_CUDA=ON`.

KrisThielemans replied:
Also, don't use `include_directories` but `target_include_directories`

Imraj-Singh replied:
done here: https://github.com/UCL/STIR/pull/1408/commits/d3f37324f65127673b04bfbaad1952c470edfab5

Imraj-Singh replied:
It's this line:
```
if (STIR_WITH_CUDA)
  target_include_directories(stir_registries PRIVATE ${CUDA_INCLUDE_DIRS})
endif()
```

KrisThielemans commented on src/cmake/STIRConfig.cmake.in:
```diff
@@ -143,6 +143,11 @@ if(@STIR_WITH_NiftyPET_PROJECTOR@)
 endif()
 
+if(@STIR_WITH_STIR_CUDA@)
+  find_package(CUDAToolkit REQUIRED)
```
possbily needs `enable_language(CUDA)` as well, not sure

Imraj-Singh replied:
done here: https://github.com/UCL/STIR/pull/1408/commits/d3f37324f65127673b04bfbaad1952c470edfab5

KrisThielemans commented on src/recon_buildblock/CUDA_stir/CudaRelativeDifferencePrior.cu:
```diff
@@ -0,0 +1,382 @@
+                                                    const int y_dim,
+                                                    const int x_dim) {
+    // tmp_grad, cp_image, cp_weights, cp_kappa, cp_penalisation_factor, cp_gamma, cp_epsilon, z_dim, y_dim, x_dim
+    int z = blockIdx.z * blockDim.z + threadIdx.z;
+    int y = blockIdx.y * blockDim.y + threadIdx.y;
+    int x = blockIdx.x * blockDim.x + threadIdx.x;
```
add `const` here (I don't think they're changing), and a comment what they are

Imraj-Singh replied:
done here: https://github.com/UCL/STIR/pull/1408/commits/f41728e6c06347fa84889cb355dcfc9a5c4d5af8

KrisThielemans commented on src/recon_buildblock/CUDA_stir/CudaRelativeDifferencePrior.cu:
```diff
@@ -0,0 +1,382 @@
+    for(int dz = min_dz; dz <= max_dz; dz++) {
+        for(int dy = min_dy; dy <= max_dy; dy++) {
+            for(int dx = min_dx; dx <= max_dx; dx++) {
+                int neighbourIndex = (z + dz) * y_dim * x_dim + (y + dy) * x_dim + (x + dx);
+                int weightsIndex = (dz + 1) * 9 + (dy + 1) * 3 + (dx + 1);
```
`const`

Imraj-Singh replied:
done here: https://github.com/UCL/STIR/commit/f41728e6c06347fa84889cb355dcfc9a5c4d5af8

KrisThielemans commented on src/recon_buildblock/CUDA_stir/CudaRelativeDifferencePrior.cu:
```diff
@@ -0,0 +1,382 @@
+            for(int dx = min_dx; dx <= max_dx; dx++) {
+                int neighbourIndex = (z + dz) * y_dim * x_dim + (y + dy) * x_dim + (x + dx);
+                int weightsIndex = (dz + 1) * 9 + (dy + 1) * 3 + (dx + 1);
+                diff = (image[inputIndex] - image[neighbourIndex]);
+                diff_abs = abs(diff);
+                add = (image[inputIndex] + image[neighbourIndex]);
+                add_3 = (image[inputIndex] + 3*image[neighbourIndex]);
```
These are all local variables, so best to decleare them here, as opposed to outside all loops. Clearer, and better for the compiler. Obviously, make them `const` as well.

KrisThielemans commented on src/recon_buildblock/CUDA_stir/CudaRelativeDifferencePrior.cu:
```diff
@@ -0,0 +1,382 @@
+
+START_NAMESPACE_STIR
+
+static void
```
let's not repeat those. We can make them static members of `RelativeDifferencePrior` or wherever they need to be

Imraj-Singh replied:
I guess this is with regards to the compute weights @KrisThielemans ?

KrisThielemans replied:
yes

KrisThielemans commented on src/recon_buildblock/CUDA_stir/CudaRelativeDifferencePrior.cu:
```diff
@@ -0,0 +1,382 @@
+    }
+
+    // Check for any errors during kernel execution
+    cudaError_t error = cudaGetLastError();
+    if (error != cudaSuccess) {
+        std::cerr << "CUDA error in compute_value kernel execution: " << cudaGetErrorString(error) << std::endl;
+        cudaFree(d_image_data); cudaFree(d_weights_data); cudaFree(d_gradient_data);
+    }
```
```suggestion
    cudaError_t cuda_error = cudaGetLastError();
    if (cuda_error != cudaSuccess) {
        cudaFree(d_image_data); cudaFree(d_weights_data); cudaFree(d_gradient_data);
        error("CUDA error in compute_value kernel execution: " << cudaGetErrorString(cuda_error));
    }
```

Imraj-Singh replied:
done here: https://github.com/UCL/STIR/pull/1408/commits/483cc138270f1f0c4922eb6bf1e48960d1a15ec6

KrisThielemans commented on src/recon_buildblock/CUDA_stir/CudaRelativeDifferencePrior.cu:
```diff
@@ -0,0 +1,382 @@
+    
+    // Copy the gradient data to the prior_gradient
+    std::copy(gradient_data.begin(), gradient_data.end(), prior_gradient.begin_all());
+    prior_gradient *= this->penalisation_factor;
```
far better to include this in the kernel

KrisThielemans commented on src/recon_buildblock/CUDA_stir/CudaRelativeDifferencePrior.cu:
```diff
@@ -0,0 +1,382 @@
+    }
+
+    // Allocate host memory for the result and copy from device to host
+    std::vector<float> gradient_data(current_image_estimate.size_all());
```
ok, but could reuse some memory (might be better to leave this though for clarity and future changes)

KrisThielemans reviewed (changes requested):
Thanks!

> There are a lot of warnings with regards to -- virtual function override intended?.

please give an example

>The set_up_cuda is a bit clunky (I did have it override the set-up previously but got a bit confused with CRTP things)

This won't help, as STIR will never call `set_up_cuda`. There's too much CRTP going on here (if any). Do you mean calling the baseclass `set_up()`?

I think it'd make more sense to use `recon_buildblock/CUDA` than `CUDA_stir` (why have stir twice?)

Imraj-Singh commented:
> > There are a lot of warnings with regards to -- virtual function override intended?.
> 
> please give an example
Looks like this:
```
/home/user/sirf/STIR/src/include/stir/Array.h(177): warning #997-D: overloaded function "stir::VectorWithOffset<T>::resize [with T=stir::Array<1, float>]" is hidden by "stir::Array<num_dimensions, elemT>::resize [with num_dimensions=2, elemT=float]" -- virtual function override intended?
          detected during:
            instantiation of class "stir::Array<num_dimensions, elemT> [with num_dimensions=2, elemT=float]" 
(109): here
            instantiation of class "stir::Array<num_dimensions, elemT> [with num_dimensions=3, elemT=float]" 
/home/user/sirf/STIR/src/include/stir/recon_buildblock/RelativeDifferencePrior.h(177): here

/home/user/sirf/STIR/src/include/stir/Array.h(180): warning #997-D: overloaded function "stir::VectorWithOffset<T>::grow [with T=stir::Array<2, float>]" is hidden by "stir::Array<num_dimensions, elemT>::grow [with num_dimensions=3, elemT=float]" -- virtual function override intended?
          detected during instantiation of class "stir::Array<num_dimensions, elemT> [with num_dimensions=3, elemT=float]" 
/home/user/sirf/STIR/src/include/stir/recon_buildblock/RelativeDifferencePrior.h(177): here

/home/user/sirf/STIR/src/include/stir/Array.h(177): warning #997-D: overloaded function "stir::VectorWithOffset<T>::resize [with T=stir::Array<2, float>]" is hidden by "stir::Array<num_dimensions, elemT>::resize [with num_dimensions=3, elemT=float]" -- virtual function override intended?
          detected during instantiation of class "stir::Array<num_dimensions, elemT> [with num_dimensions=3, elemT=float]" 
/home/user/sirf/STIR/src/include/stir/recon_buildblock/RelativeDifferencePrior.h(177): here

/home/user/sirf/STIR/src/include/stir/recon_buildblock/RelativeDifferencePrior.h(152): warning #997-D: function "stir::GeneralisedPrior<DataT>::set_up(const std::shared_ptr<const DataT> &) [with DataT=stir::DiscretisedDensity<3, float>]" is hidden by "stir::RelativeDifferencePrior<elemT>::set_up [with elemT=float]" -- virtual function override intended?
          detected during:
            instantiation of class "stir::RelativeDifferencePrior<elemT> [with elemT=float]" 
/home/user/sirf/STIR/src/include/stir/recon_buildblock/CUDA/CudaRelativeDifferencePrior.h(12): here
            instantiation of class "stir::CudaRelativeDifferencePrior<elemT> [with elemT=float]" 
/home/user/sirf/STIR/src/recon_buildblock/CUDA/CudaRelativeDifferencePrior.cu(376): here
```

KrisThielemans commented:
ok. Nothing to do with this PR. Would you mind creating a separate issue for it? Please add which compiler you're using. (It's a false positive, but would be nice to have a clean way to prevent the warning)

KrisThielemans commented on CMakeLists.txt:
```diff
@@ -186,6 +187,24 @@ else()
+if(NOT DISABLE_STIR_CUDA)
+  ## check 
+  find_package(CUDAToolkit REQUIRED)
+  find_package(CUDA REQUIRED)
```
are you sure this is ok? I think this is outdated and `check_language` should be used

KrisThielemans commented on src/CMakeLists.txt:
```diff
@@ -244,6 +244,9 @@ add_library(stir_registries OBJECT ${STIR_REGISTRIES})
 target_include_directories(stir_registries PRIVATE ${STIR_INCLUDE_DIR})
 target_include_directories(stir_registries PRIVATE ${Boost_INCLUDE_DIR})
+if (STIR_WITH_CUDA)
+  target_include_directories(stir_registries PRIVATE ${CUDA_INCLUDE_DIRS})
```
I really hope we don't need `CUDA_INCLUDE_DIRS` for the registries. STIR .h files ideally have no CUDA includes at all. It's going to create trouble.

Imraj-Singh replied:
I thought we'd want the target to be the registeries as we need the cuda_runtime.h when building cuda files that will be in the registry... Also I couldn't quite figure out where else to put the target_include_directories, where would you suggest?

KrisThielemans replied:
As far as I can see,`cuda_runtime.h should NOT be included in src/include/stir/recon_buildblock/CUDA/CudaRelativeDifferencePrior.h, but in the .cxx/.cu. That means that the registries wouldn't need to know about the CUDA include dir, therefore you don't need the `target_include_directories` at all. (The .cu files need it, but they get it by depending on CUDA targets).

KrisThielemans commented on src/include/stir/recon_buildblock/CUDA/CudaRelativeDifferencePrior.h:
```diff
@@ -0,0 +1,36 @@
+
+#include "stir/recon_buildblock/RelativeDifferencePrior.h"
+#include "stir/DiscretisedDensity.h"
+#include <cuda_runtime.h>
```
move to cxx

Imraj-Singh replied:
this is done but dim3 is cuda specific, I changed it for a native struct and removed the cuda_runtime include, aren't the other two needed?

KrisThielemans commented:
> > ok. Nothing to do with this PR. Would you mind creating a separate issue for it? Please add which compiler you're using. (It's a false positive, but would be nice to have a clean way to prevent the warning)
> 
> I am having some issues getting it to register correctly in the stir registries. Perhaps it's related to that? 

no

> I am not sure I am correctly making `CudaRelativeDifferencePrior` a child of `RelativeDifferencePrior`. I am a bit hesistant to open a issue as I think something is a bit wrong at the moment.

Please show errors/problems. I have no time to try this myself. sorry.

Imraj-Singh commented:
> > > ok. Nothing to do with this PR. Would you mind creating a separate issue for it? Please add which compiler you're using. (It's a false positive, but would be nice to have a clean way to prevent the warning)
> > 
> > 
> > I am having some issues getting it to register correctly in the stir registries. Perhaps it's related to that?
> 
> no
> 
> > I am not sure I am correctly making `CudaRelativeDifferencePrior` a child of `RelativeDifferencePrior`. I am a bit hesistant to open a issue as I think something is a bit wrong at the moment.
> 
> Please show errors/problems. I have no time to try this myself. sorry.

Thanks @KrisThielemans 

I was running `./STIR/build/src/utilities/stir_list_registries`  and did not see the prior under `GeneralisedPrior<DiscretisedDensity<3,float>>`. The output looks like this:

```
WARNING: FactoryRegistry:: overwriting previous value of key in registry.
     key: None
------------ ProjectorByBinPair --------------
Matrix
None
Parallelproj
Separate Projectors
------------ ForwardProjectorByBin --------------
Matrix
None
Parallelpr etc etc
```

Is the overwriting causing the issue, or I am not sure if I have updated all the neccessary files to register it correctly. I guess part of my concern is whether I updated all the neccessary files so that STIR can include the new class in the registeries. This is a quite new and pretty confusing

KrisThielemans commented:
Do you have the equivalent of https://github.com/UCL/STIR/blob/8ced2d73933420457e0bc76074964b7d4ff00f0c/src/recon_buildblock/RelativeDifferencePrior.cxx#L140C1-L141C97
Maybe these days this can be done in the .h file (which would be cleanest)

Imraj-Singh commented on src/recon_buildblock/CUDA/CudaRelativeDifferencePrior.cu:
```diff
@@ -0,0 +1,384 @@
+// Explicit template instantiations
+template class stir::CudaRelativeDifferencePrior<float>;
+template <typename elemT>
+const char* const CudaRelativeDifferencePrior<elemT>::registered_name = "Cuda Relative Difference Prior";
```
@KrisThielemans yes I added an overwrite it

KrisThielemans replied:
please use exact same syntax as in the lines quoted above. C++ is a bit tricky there. (or try and see what happens if you put the initialiser in the .h file). If that doesn't work, comment out the one in `RelativeDifferencePrior`, or remove that from the registry completely. I don't think we've ever tried deriving from an existing class in the registry.

Imraj-Singh replied:
@KrisThielemans is the reason for not deriving from an existing class in the registry, because it's just a bad idea? I couldn't think of a better idea...

KrisThielemans replied:
Not sure what you mean. Ideally we can derive from an existing class, but if it doesn't work, we'll have to find another way...

So first thing to do is to fix your syntax and see what happens.

KrisThielemans commented:
@Imraj-Singh any chance you can pick this up? I think my suggestions might resolve your trouble. Also, it'll need a merge with `master` of course.

I'd like to you the main CUDA support stuff elsewhere...

Imraj-Singh commented:
> @Imraj-Singh any chance you can pick this up? I think my suggestions might resolve your trouble. Also, it'll need a merge with `master` of course.
> 
> I'd like to you the main CUDA support stuff elsewhere...

Just back I will spend some time the next couple of days.

Imraj-Singh commented:
@KrisThielemans I tried removing the registering of `RelativeDifferencePrior` but still having errors. FYI I took all the content out of the CUDA files to try and diagnose the issue.

KrisThielemans commented:
> /usr/bin/ld: ../recon_buildblock/librecon_buildblock.a(CudaRelativeDifferencePrior.cu.o):(.data.rel.ro.local+0x0): multiple definition of stir::CudaRelativeDifferencePrior::registered_name;

This is because your initialisation of `registered_name` is in the .h file. That means it will exist in every .cxx/.cu file that includes, so there are multiple definitions.

Possible solutions:
- move it to the .cu
- move it to a new .cxx
- try to initialise inside the class definition, not outside. This apparently needs C++-17 https://en.cppreference.com/w/cpp/language/static, but we've moved to that anyway (you might have to merge master)
  ```c++
  class CudaRelativeDifferencePrior : public RelativeDifferencePrior<elemT> {
    public:
        using RelativeDifferencePrior<elemT>::RelativeDifferencePrior;
        // Name which will be used when parsing a GeneralisedPrior object
        inline static const char* const registered_name = "CudaRelativeDifferencePrior;
       ...
   ```

If the latter works, it's my preferred solution (by far clearest for the user), and we could do it for the rest of STIR now as well.

KrisThielemans commented:
Also, please add `check_language(CUDA)` and if it isn't there, set `DISABLE_STIR_CUDA=ON`, possibly with a warning. At present, if you don't have CUDA, it just fails without any help what to do.

KrisThielemans commented:
@Imraj-Singh this all works now. Changes needed were quite small. Unfortunately, I had to run `dos2unix` as well as `clang-format`. That means you might want to do `git diff -w` to ignore white space. Also, it will let git attribute all these lines to me, most likely. Sorry.

Question is then if this is production ready... what do you think?

(we should convert the test to a unit test)

KrisThielemans commented:
All tests on GHA complain about the "import" of the base-class constructors in `RegisteredParsingObject`. I don't know why. It works fine in my set-up (gcc 12.3)

KrisThielemans commented:
Ah, the errors are due to SWIG.
https://github.com/UCL/STIR/actions/runs/9800995632/job/27063623699?pr=1408#step:10:120
```
/home/runner/work/STIR/STIR/src/include/stir/RegisteredParsingObject.h:80: Error: Declaration of 'Parent' shadows template parameter,
/home/runner/work/STIR/STIR/src/include/stir/RegisteredParsingObject.h:78: Error: previous template parameter declaration 'Parent'.
```

KrisThielemans commented:
This should be nearly done now. For a 256x251x211 image with 2 cylinders, `cuda_rdp_tests` gives
```
norms: diff of gradients: 5.09262e-07 org: 64.8277 rel: 7.85563e-09
The Prior Value (CUDA) = 3385.94
The Prior Value (CPU) = 3385.94
Difference = 3.06435e-05
Value CUDA time = 1.35823s
Value CPU time = 8.6639s
Gradient CUDA time = 0.555103s
Gradient CPU time = 12.3153s
```
It also works for a smaller image with much more structure.

Sadly, I have problems in `test_priors`.
```
----- test Cuda Relative Difference Prior_no_kappa_with_eps  --> Gradient

INFO: Computing gradient

INFO: Computing objective function at target

INFO: Computing gradient of objective function by numerical differences (this will take a while)
Error : unequal values are 1.71268 and 1.61074. gradient
Numerical gradient test failed with for Cuda Relative Difference Prior_no_kappa_with_eps
Writing diagnostic files Cuda Relative Difference Prior_no_kappa_with_eps_target.hv, *gradient.hv (and *numerical_gradient.hv if full gradient test is used)
----- test Cuda Relative Difference Prior_no_kappa_with_eps  --> Hessian-vector product for convexity
----- test Cuda Relative Difference Prior_no_kappa_with_eps  --> Hessian against numerical
Error : unequal values are 1.19587 and 1.21844. Hessian
Numerical-Hessian test failed with for Cuda Relative Difference Prior_no_kappa_with_eps prior
----- testing Hessian-vector product (accumulate_Hessian_times_input)

INFO: Comparing Hessian*dx with difference of gradients

INFO: Computing gradient

ERROR: CUDA error in compute_gradient kernel execution: invalid argument
terminate called after throwing an instance of 'std::runtime_error'
  what():
ERROR: CUDA error in compute_gradient kernel execution: invalid argument

Aborted
```
Doing a visual check on the analytic vs numerical gradient indicates they are very similar, but numerically the diff is rather large.
![image](https://github.com/UCL/STIR/assets/6362141/cbb55d88-e80c-44ad-bd65-23276a21f0d6)
(left images have range ~[-15.5,5.6] while the diff on the right has range ~[-.5,5])

The kernel invalid argument is in 
```
compute-sanitizer --tool memcheck src/recon_test/test_priors
...
========= Program hit cudaErrorInvalidValue (error 1) due to "invalid argument" on CUDA API call to cudaMemcpy.
=========     Saved host backtrace up to driver entry point at error
=========     Host Frame: [0x34d163]
=========                in /usr/lib/wsl/drivers/nvdd.inf_amd64_fa2520b946613838/libcuda.so.1.1
=========     Host Frame:cudaMemcpy [0x6b8aa]
=========                in /usr/local/cuda-12.2/targets/x86_64-linux/lib/libcudart.so.12
=========     Host Frame:void stir::array_to_device<3, float>(float*, stir::Array<3, float> const&) [0x28b8a4]
=========                in /home/kris/devel/buildCondaGT/builds/STIR/build/src/recon_test/test_priors
=========     Host Frame:stir::CudaRelativeDifferencePrior<float>::compute_gradient(stir::DiscretisedDensity<3, float>&, stir::DiscretisedDensity<3, float> const&) [0x28bab3]
```
I don't know why.

KrisThielemans commented:
Note that the test image are very small (8x9x10). Not sure if that matters or not. @Imraj-Singh ?

KrisThielemans commented:
The above problems have now been resolved. 2 things remaining
- the numerical Hessian test for the RDP tails with kappa (independent of the CUDA stuff)
- GHA cannot run CUDA, so we have to disable the tests for CudaRDP in test_priors if there's no CUDA run-time

KrisThielemans commented:
Everything fine now, except that there's no CUDA run-time on GHA, so we have to disable that test somehow.

KrisThielemans commented:
All good now! (except a Zenodo time-out)

KrisThielemans commented:
Should be done now. I've added the timings to `stir_timings`. With default mMR image size, I get on my desktop (AMD Ryzen 9 5900 12-Core, RTX 3070, default num_threads=21) (timings in ms, first is CPU-time, second is wall-clock time)
```
        RDP_value                                                637.000                         524.449
        RDP_grad                                                4607.667                         879.874
        Cuda_RDP_value                                          1732.778                          86.533
        Cuda_RDP_grad                                            919.778                          46.042
```
compared to parallelproj forward proj 225ms and back proj 573ms.

Thanks @Imraj-Singh!

KrisThielemans reviewed (approved):
All done
